src/ctapipe/core/tool.py

```diff
@@ -425,26 +426,44 @@ def run(self, argv=None, raises=False):
 
                 self.start()
                 self.finish()
-                self.log.info(f"Finished: {self.name}")
+                self.log.info("Finished: %s", self.name)
                 Provenance().finish_activity(activity_name=self.name)
             except (ToolConfigurationError, TraitError) as err:
                 self.log.error("%s", err)
                 self.log.error("Use --help for more info")
                 exit_status = 2  # wrong cmd line parameter
+                Provenance().finish_activity(
+                    activity_name=self.name, status="error", exit_code=exit_status
+                )
                 if raises:
                     raise
             except KeyboardInterrupt:
                 self.log.warning("WAS INTERRUPTED BY CTRL-C")
+                exit_status = 130  # Script terminated by Control-C
                 Provenance().finish_activity(
-                    activity_name=self.name, status="interrupted"
+                    activity_name=self.name, status="interrupted", exit_code=exit_status
                 )
-                exit_status = 130  # Script terminated by Control-C
             except Exception as err:
-                self.log.exception(f"Caught unexpected exception: {err}")
-                Provenance().finish_activity(activity_name=self.name, status="error")
+                self.log.exception("Caught unexpected exception: %s", err)
                 exit_status = 1  # any other error
+                Provenance().finish_activity(
+                    activity_name=self.name, status="error", exit_code=exit_status
+                )
                 if raises:
                     raise
+            except SystemExit as err:
+                if raises:
+                    raise  # do not re-intercept in tests
+                else:
+                    exit_status = err.code
+                    self.log.exception(
+                        "Caught SystemExit with exit code %s", exit_status
+                    )
+                    Provenance().finish_activity(
+                        activity_name=self.name,
+                        status="partial_success",
```

Comment:
Why is catching a non-zero exit code here a "partial_success" and not just "error"

I don't really understand why we here need "partial_success" and it's very ambiguous. I think we should just use error here. Any semantic meaning can be then assigned downstream to the individual error codes.

Reply:
I agree that "partial_success" is ambiguous, on the other hand, my point was to differentiate from the UnknownError condition. However, I tend to agree with you that all three conditions outlined in exceptions (TraitError, UnknownError and SystemExit) are all "errors" in general meaning and the further disentanglement can (and probably should) be done by the error codes.